workflow/vm/vm.go

```diff
@@ -0,0 +1,164 @@
+			// chain has completed and it can be removed.
+			if c.IsDone() {
+				v.executionChains = append(v.executionChains[:i], v.executionChains[i+1:]...)
+			}
+			if ok {
+				v.ExecuctionRequests <- execReq
+				break
+			}
+		}
+	}
+	// event can also be a start of some new lifecycles for many workflows.
+	// create a full trigger spec for the event and find out the matching workflows.
+	t := e.createTrigger()
+	for _, w := range v.workflows {
+		// check if workflow matches with trigger created from event.
+		if ok := w.Match(t); ok {
+			// create a new chain, register it and create task execution request for
+			// its first task.
+			c := newChain(w)
+			ok, execReq := c.Next(e)
+			// chain might have just one task, if so, don't bother to add it chain list.
+			if !c.IsDone() {
+				v.executionChains = append(v.executionChains, c)
+			}
+			// check against empty workflow Tasks.
+			if ok {
+				v.ExecuctionRequests <- execReq
+			}
+		}
+	}
+}
+
+// Add adds a workflow to VM.
+func (v *VM) Add(w workflow.Workflow) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	v.workflows = append(v.workflows, w)
+}
+
+// Remove removes workflow and all of its execution chains from VM.
+func (v *VM) Remove(hash string) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	for i, w := range v.workflows {
+		if w.Hash == hash {
+			v.workflows = append(v.workflows[:i], v.workflows[i+1:]...)
+			for ii := len(v.executionChains) - 1; i >= 0; i-- {
+				if v.executionChains[ii].w.Hash == hash {
+					v.executionChains = append(v.executionChains[:ii], v.executionChains[ii+1:]...)
+				}
+			}
+			return
+		}
+	}
+}
+
+// Shutdown immediately ends vm and closes ExecutionRequests channel.
+func (v *VM) Shutdown() {
+	v.m.Lock()
+	defer v.m.Unlock()
+	v.closing = true
+	close(v.ExecuctionRequests)
+}
+
+// Event represents a new event/result input to the workflow VM.
+type Event struct {
+	// InstanceHash that event is emitted from.
+	InstanceHash string
+
+	// ParentHash of event if there is.
+	ParentHash []byte
+
+	// Key of event.
+	Key string
+
+	// TaskKey of event when event is emitted as a response to an execution.
+	TaskKey string
+
+	// Data is the payload of event.
+	Data map[string]interface{}
+
+	// Secret is generated by the Workflow VM for each execution request in a chain to check which
+	// Event belongs to which Execution request's result.
+	Secret string
+}
+
+// createTrigger creates a new trigger schema from event which is used to match the
+// event with registered workflows in VM.
+func (e Event) createTrigger() workflow.Trigger {
+	return workflow.Trigger{
+		EventKey:     e.Key,
+		InstanceHash: e.InstanceHash,
+		Filter: workflow.Filter{
+			TaskKey: e.TaskKey,
+		},
+	}
+}
+
+// Execution is an execution request that should be performed by the caller.
+// Each received Event from the Process() may cause multiple Execution requests
+// to be emitted by the Workflow VM.
+type Execution struct {
```

Comment:
Why Execution is redefined here? Shound't we use `execution.Execution` ?

Reply:
this is not same as `execution.Execution`. this is actually an execution request. i may change the name to `ExecutionRequest`.

also, we cannot create a new `execution.Execution` in `workflowvm` because creation should be made one by the `execution.Create()` func which is actually called by the `sdk.ExecuteTask()`. so if you want to reuse execute task sdk, you have to pass the params it needs not `execution.Execution`. It is also nice to have `workflowvm` not fully aware of these types for simplicity.